nni/retiarii/trainer/pytorch/base.py

```diff
@@ -188,57 +213,64 @@ def __init__(self, multi_model, kwargs=[]):
         for _ in range(max_epochs):
             self._train()
+        nni.report_final_result(self._validate())
 
     def _train(self):
-        for batch_idx, multi_model_batch in enumerate(zip(*self._dataloaders)):
+        for batch_idx, multi_model_batch in enumerate(zip(*self._train_dataloaders)):
             for opt in self._optimizers:
                 opt.zero_grad()
             xs = []
             ys = []
             for idx, batch in enumerate(multi_model_batch):
-                x, y = self.training_step_before_model(batch, batch_idx, f'cuda:{idx}')
+                x, y = self.training_step_before_model(
+                    batch, batch_idx, f'cuda:{idx}')
                 xs.append(x)
                 ys.append(y)
-            
+
             y_hats = self.multi_model(*xs)
             if len(ys) != len(xs):
                 raise ValueError('len(ys) should be equal to len(xs)')
             losses = []
             report_loss = {}
             for output_idx, yhat in enumerate(y_hats):
                 if len(ys) == len(y_hats):
-                    loss = self.training_step_after_model(xs[output_idx], ys[output_idx], yhat)
+                    loss = self.training_step_after_model(
+                        xs[output_idx], ys[output_idx], yhat)
                 elif len(ys) == 1:
-                    loss = self.training_step_after_model(xs[0], ys[0].to(yhat.get_device()), yhat)
+                    loss = self.training_step_after_model(
+                        xs[0], ys[0].to(yhat.get_device()), yhat)
                 else:
-                    raise ValueError('len(ys) should be either 1 or len(y_hats)')
+                    raise ValueError(
+                        'len(ys) should be either 1 or len(y_hats)')
                 losses.append(loss.to("cuda:0"))
-                report_loss[self.kwargs['model_kwargs'][output_idx]['model_id']] = loss.item()
+                report_loss[self.kwargs['model_kwargs']
+                            [output_idx]['model_id']] = loss.item()
             summed_loss = sum(losses)
             summed_loss.backward()
             for opt in self._optimizers:
                 opt.step()
-            if batch_idx % 50 == 0:
-                nni.report_intermediate_result(report_loss)
+            # if batch_idx % 50 == 0:
+            #     nni.report_intermediate_result(report_loss)
             if self.max_steps and batch_idx >= self.max_steps:
                 return
 
-    
     def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor], batch_idx: int) -> Dict[str, Any]:
         x, y = self.training_step_before_model(batch, batch_idx)
         y_hat = self.model(x)
         return self.training_step_after_model(x, y, y_hat)
 
-    def training_step_before_model(self, batch: Tuple[torch.Tensor, torch.Tensor], batch_idx: int, device = None):
+    def training_step_before_model(self, batch: Tuple[torch.Tensor, torch.Tensor], batch_idx: int, device=None):
```

Comment:
Suggest using `self.device`

Reply:
In MultiModel, different model's input may need to be placed on different devices (called in _train). Currently, the trainer just sets one GPU per model in hard-code.

BTW, train_step and validation_step are not used in PyTorchImageClassificationTrainer. Removed.